### DISCUSSION TITLE
How to extend Airflow image to configure Spark сonnection right?

### DISCUSSION THREAD

**Discussion by makrushin-evgenii** [apache/airflow#18138] [Q&A] [answered]
Do I understand correctly that for the SparkSubmitOperator to work, Spark-binaries must be added to the image, which means that Java and Hadoop also must be configured? Can you clarify the requirements to configure [Apache Spark Connection](https://airflow.apache.org/docs/apache-airflow-providers-apache-spark/stable/connections/spark.html#apache-spark-connection)?

I want to continue using the [docker-compose.yaml](https://airflow.apache.org/docs/apache-airflow/stable/start/docker.html#docker-compose-yaml) from the "Quick Start / Running Airflow in Docker" page with minimal changes. Perhaps there are good ready-made examples of such an image extending?

My project now looks like this:
```
# docker-compose.yaml
...
x-airflow-common:
  &airflow-common
  build: .
...

# Dockerfile
FROM apache/airflow:2.1.3

COPY requirements.txt .
RUN pip install -r requirements.txt

# requirements.txt
apache-airflow-providers-apache-spark==2.0.0
```

Applications in our YARN cluster are submitted from cmd like this:
```
./bin/spark-submit.cmd `
  --master yarn `
  --deploy-mode cluster `
  --keytab C:\Tools\spark-3.0.1-bin-hadoop3.2\keytabs\SrsApp@DEV.KONTUR.RU.keytab  `
  --principal SrsApp@DEV.KONTUR.RU `
  examples/src/main/python/pi.py 10
```

**Comment by potiuk** [comment]
We have https://airflow.apache.org/docs/docker-stack/recipes.html - with Hadoop installation which includes Java so, likely you are on a good start. 

I would really love if - when you get to the right "Recipe" contribute it back to the very documentation. Airflow is developed by the community so such contributions are most welcome - and it paves the way for other users.  - it's super easy  - just click on "Suggest Improvement on this Page" button and you will have a PR opened that you will be able to submit easily. 

I'd love to get such tested and proven recipe for spark in our docs.

**Reply by potiuk** [reply]
And just a warning (also for others looking here in the future) as you likely noticed, the docker compose is "quick-start" not "production ready" so likely you will have to modify it if want persistence, security, reliability, resilience etc. If you use it for production, there is likely quite some work to make it "production-ready". 

Instead, the Helm Chart is something that we target as production-ready: https://airflow.apache.org/docs/helm-chart/stable/index.html

**Reply by makrushin-evgenii** [reply]
> We have https://airflow.apache.org/docs/docker-stack/recipes.html - with Hadoop installation which includes Java so, likely you are on a good start.

Thanks for the answer, this link helped a lot

> I would really love if - when you get to the right "Recipe" contribute it back to the very documentation. 

Good. I will definitely try

**Comment by ufo108** [comment, marked as answer]
> We have https://airflow.apache.org/docs/docker-stack/recipes.html - with Hadoop installation which includes Java so, likely you are on a good start.

@potiuk thank you for the useful link.

@makrushin-evgenii  building image at https://airflow.apache.org/docs/docker-stack/recipes.html  , you can add the spark section like this:
```dockerfile
# after HADOOP section
ENV HADOOP_HOME /opt/hadoop
ENV HADOOP_CLASSPATH $($HADOOP_HOME/bin/hadoop classpath)
ENV HADOOP_CLASSPATH $HADOOP_CLASSPATH:/opt/libs/*
ENV HADOOP_CONF_DIR=$HADOOP_HOME/etc/hadoop
ENV PATH $PATH:$HADOOP_HOME/bin

# SPARK
ENV SPARK_VERSION 3.0.0
RUN wget --quiet -O ./spark.tgz https://archive.apache.org/dist/spark/spark-$SPARK_VERSION/spark-$SPARK_VERSION-bin-hadoop3.2.tgz && \
    tar -xzf ./spark.tgz && \
    rm ./spark.tgz && \
    ln -s /home/sparker/spark-$SPARK_VERSION-bin-hadoop3.2 /opt/spark

ENV SPARK_HOME /opt/spark
ENV SPARK_DIST_CLASSPATH=$HADOOP_CLASSPATH:$SPARK_HOME/jars/*
ENV PATH $PATH:${SPARK_HOME}/bin:${SPARK_HOME}/sbin
```

Moreover, you can describe in advance the required configuration of the HADOOP components in the files:
- yarn-site.xml
- fair-scheduler.xml
- core-site.xml
- hive-site.xml
- spark-defaults.conf

by placing them in the image as follows:
```dockerfile
# CONFIGS
COPY ./config/yarn-site.xml $HADOOP_HOME/etc/hadoop
COPY ./config/fair-scheduler.xml $HADOOP_HOME/etc/hadoop
COPY ./config/core-site.xml $HADOOP_HOME/etc/hadoop
COPY ./config/hive-site.xml $HADOOP_HOME/etc/hadoop
COPY ./config/spark-defaults.conf $SPARK_HOME/conf
```

The only major painful issue is the order in which Spark uses the libraries.

For example, I tried to start streaming processing through `PySpark`, but the code crashed with the exception of the absence of some method in the `Kafka client` library.

Later I found out that `Spark` was using the old version of the `Kafka client` library from the Hadoop libraries (Although the `Spark` includes the newer version of `Kafka client` library).  In my case, the order of the libraries in the **HADOOP_CLASSPATH** environment variable included first `HADOOP` and then `Spark` path to libraries.

I managed to solve the problem by explicitly setting the order of using the libraries through **spark.driver.userClassPathFirst** variable as follows:
```python
import os
from pyspark.sql import SparkSession

host_ip = os.getenv('HOST_IP')
username = os.getenv('USERNAME')
print(host_ip, username)


spark_jar_packages = ','.join([
    'org.apache.spark:spark-sql-kafka-0-10_2.12:3.0.0',
    'org.apache.spark:spark-avro_2.12:3.0.0'
])


spark = SparkSession \
    .builder \
    .config("spark.driver.host", host_ip) \
    .config("spark.driver.bindAddress", '0.0.0.0') \
    .config("spark.driver.port", 33139) \
    .config("spark.driver.blockManager.port", 45029) \
    .config('spark.jars.packages', spark_jar_packages) \
    .config("spark.driver.userClassPathFirst", True) \
    .config("spark.jars.excludes", "org.slf4j:*") \
    .appName(f"KafkaStreaming2_{host_ip}-{username}") \
    .enableHiveSupport() \
    .getOrCreate()
```

**Reply by makrushin-evgenii** [reply]
@Melchizedek13, for some reason there is no "mark as answer" button (UPD: already marked)

It was enough for me to copy the Java installation section from this recipe. I specified the `spark-3.0.1-bin-hadoop3.2` directory as `HADOOP_HOME` and that was enough for spark-submit to work:
```dockerfile
FROM apache/airflow:2.1.3

# Install Python requirements
COPY requirements.txt .
RUN pip install -r requirements.txt

USER 0
SHELL ["/bin/bash", "-o", "pipefail", "-e", "-u", "-x", "-c"]

# Install Java
RUN mkdir -pv /usr/share/man/man1 \
    && mkdir -pv /usr/share/man/man7 \
    && curl -fsSL https://adoptopenjdk.jfrog.io/adoptopenjdk/api/gpg/key/public | apt-key add - \
    && echo 'deb https://adoptopenjdk.jfrog.io/adoptopenjdk/deb/ buster main' > \
        /etc/apt/sources.list.d/adoptopenjdk.list \
    && apt-get update \
    && apt-get install --no-install-recommends -y \
      adoptopenjdk-8-hotspot-jre \
    && apt-get autoremove -yqq --purge \
    && apt-get clean \
    && rm -rf /var/lib/apt/lists/*
ENV JAVA_HOME=/usr/lib/jvm/adoptopenjdk-8-hotspot-jre-amd64

# Install Spark
ENV SPARK_HOME=/spark-3.0.1-bin-hadoop3.2
ENV HADOOP_HOME=$SPARK_HOME
ENV HADOOP_CONF_DIR=$SPARK_HOME/conf

RUN curl -O https://archive.apache.org/dist/spark/spark-3.0.1/spark-3.0.1-bin-hadoop3.2.tgz \
    && tar -C / -zxvf ./spark-3.0.1-bin-hadoop3.2.tgz

COPY resources/spark_config $HADOOP_CONF_DIR
```

About @potiuk request:
> I'd love to get such tested and proven recipe for spark in our docs.

I got a Windows Docker-specific error that caused the execution of spark-submit to stop at the "looking for the active RM" step. Therefore, I cannot say that I tested this solution to supplement the documentation with it
```
[2021-09-15 12:34:38,363] {spark_submit.py:526} INFO - 21/09/15 12:34:38 INFO client.RequestHedgingRMFailoverProxyProvider: Looking for the active RM in [rm1, rm2]...
```
